Userland/Demos/CatDog/CatDog.cpp

```diff
@@ -10,174 +10,187 @@
-    catdog->m_serun2 = *TRY(Gfx::Bitmap::try_load_from_file("/res/icons/catdog/serun2.png"sv));
-    catdog->m_sleep1 = *TRY(Gfx::Bitmap::try_load_from_file("/res/icons/catdog/sleep1.png"sv));
-    catdog->m_sleep2 = *TRY(Gfx::Bitmap::try_load_from_file("/res/icons/catdog/sleep2.png"sv));
-    catdog->m_srun1 = *TRY(Gfx::Bitmap::try_load_from_file("/res/icons/catdog/srun1.png"sv));
-    catdog->m_srun2 = *TRY(Gfx::Bitmap::try_load_from_file("/res/icons/catdog/srun2.png"sv));
-    catdog->m_still = *TRY(Gfx::Bitmap::try_load_from_file("/res/icons/catdog/still.png"sv));
-    catdog->m_swrun1 = *TRY(Gfx::Bitmap::try_load_from_file("/res/icons/catdog/swrun1.png"sv));
-    catdog->m_swrun2 = *TRY(Gfx::Bitmap::try_load_from_file("/res/icons/catdog/swrun2.png"sv));
-    catdog->m_wrun1 = *TRY(Gfx::Bitmap::try_load_from_file("/res/icons/catdog/wrun1.png"sv));
-    catdog->m_wrun2 = *TRY(Gfx::Bitmap::try_load_from_file("/res/icons/catdog/wrun2.png"sv));
-    catdog->m_curr_bmp = catdog->m_alert;
+    for (auto const& image_source : image_sources)
+        catdog->m_images.append({ image_source.state, *TRY(Gfx::Bitmap::try_load_from_file(image_source.path)) });
+
     return catdog;
 }
 
-void CatDog::timer_event(Core::TimerEvent&)
+CatDog::CatDog()
+    : m_proc_all(MUST(Core::Stream::File::open("/sys/kernel/processes"sv, Core::Stream::OpenMode::Read)))
+    , m_state { State::ALERT }
+{
+    m_idle_sleep_timer.start();
+}
+
+void CatDog::set_roaming(bool roaming)
+{
+    if (!roaming && (m_state & SPECIAL_IDLE_MASK) == 0)
+        m_state = State::ALERT;
+    m_state &= ~State::SLEEPING;
+    m_roaming = roaming;
+    update();
+}
+
+uint16_t CatDog::special_application_states() const
 {
     auto maybe_proc_info = Core::ProcessStatisticsReader::get_all(*m_proc_all);
-    if (!maybe_proc_info.is_error()) {
-        auto proc_info = maybe_proc_info.release_value();
-
-        auto maybe_paint_program = proc_info.processes.first_matching([](auto& process) {
-            return process.name.equals_ignoring_case("pixelpaint"sv) || process.name.equals_ignoring_case("fonteditor"sv);
-        });
-        if (maybe_paint_program.has_value()) {
-            m_main_state = MainState::Artist;
-        } else {
-            auto maybe_inspector_program = proc_info.processes.first_matching([](auto& process) {
-                return process.name.equals_ignoring_case("inspector"sv) || process.name.equals_ignoring_case("systemmonitor"sv) || process.name.equals_ignoring_case("profiler"sv);
-            });
-
-            if (maybe_inspector_program.has_value())
-                m_main_state = MainState::Inspector;
-            // If we currently have an application state but that app isn't open anymore, go back to idle.
-            else if (!is_non_application_state(m_main_state))
-                m_main_state = MainState::Idle;
-        }
-    }
+    if (maybe_proc_info.is_error())
+        return {};
```

Comment:
Semantically this means `0`, which is `State::ALERT`, right? That's a bit odd in the context of this function alone. Should `0` be a "no state" or "unknown" value?

Doing that would break line 152 where clearing idle puts us in alert automatically though.

Reply:
Yes, overloading 0 was a bad idea. Fixed.